packages/ec-core/src/operator/selector/tournament.rs

```diff
@@ -26,15 +39,91 @@ where
         rng: &mut ThreadRng,
     ) -> Result<&'pop P::Individual> {
         ensure!(
-            population.size() >= self.size,
+            population.size() >= self.size.into(),
             "The population had size {} and we wanted a tournament of size {}",
             population.size(),
             self.size
         );
         population
             .as_ref()
-            .choose_multiple(rng, self.size)
+            .choose_multiple(rng, self.size.into())
             .max()
-            .with_context(|| "The tournament was empty; should have been {size}")
+            .with_context(|| format!("The tournament was empty; should have been {}", self.size))
+    }
+}
+
+#[cfg(test)]
+#[rustversion::attr(before(1.81), allow(clippy::unwrap_used))]
+#[rustversion::attr(
+    since(1.81),
+    expect(
+        clippy::unwrap_used,
+        reason = "`max()` can fail if the list of individuals is empty, but we know that can't \
+                  happen so we'll unwrap"
+    )
+)]
+#[rustversion::attr(before(1.81), allow(clippy::arithmetic_side_effects))]
+#[rustversion::attr(
+    since(1.81),
+    expect(
+        clippy::arithmetic_side_effects,
+        reason = "The tradeoff safety <> ease of writing arguably lies on the ease of writing \
+                  side for test code."
+    )
+)]
+mod tests {
+    use std::num::NonZeroUsize;
+
+    use rand::thread_rng;
+    use test_strategy::proptest;
+
+    use super::Tournament;
+    use crate::{individual::ec::EcIndividual, operator::selector::Selector};
+
+    #[test]
+    fn tournament_size_1() {
+        let mut rng = thread_rng();
+        let scores = &[5, 8, 9];
+        let population = scores
+            .iter()
+            .enumerate()
+            .map(|(genome, score)| EcIndividual::new(genome, score))
+            .collect::<Vec<_>>();
+        let selector = Tournament::new(NonZeroUsize::MIN);
+        let winner = selector.select(&population, &mut rng).unwrap();
+        assert!(scores.contains(winner.test_results));
+    }
+
+    #[proptest]
+    fn tournament_size_2_pop_size_2(#[any] x: i32, #[any] y: i32) {
+        let mut rng = thread_rng();
+        let scores = &[x, y];
+        let population = scores
+            .iter()
+            .enumerate()
+            .map(|(genome, score)| EcIndividual::new(genome, score))
+            .collect::<Vec<_>>();
+        let selector = Tournament::binary();
+        let winner = selector.select(&population, &mut rng).unwrap();
+        assert_eq!(winner.test_results, &x.max(y));
+    }
+
+    #[proptest]
+    fn tournament_size_2_pop_size_3(
+        #[strategy(-1000..1000)] x: i32,
+        #[strategy(-1000..1000)] y: i32,
+        #[strategy(-1000..1000)] z: i32,
+    ) {
+        let mut rng = thread_rng();
+        let scores = &[3 * x, 3 * y + 1, 3 * z + 2];
```

Comment:
This just ensures that all three values are distinct, which allows me to use use `>` (instead of `>=`) in the assertion below.

Reply:
just do `tournament_size_2_pop_size_3(#[any] a: i32, #[filter(|x| x != #a)] b: i32, #[filter(|x| x != a && x != b)] c: i32)` - yes `filter` isn't the best thing ever but since this case is so rare it shouldn't really matter in practice.

Reply:
I could go that route, but my sense from reading the `proptest` book in various airports (and with varying amounts of sleep deprivation!) is that it really works "with" `proptest` if you avoid `filter` and that you're better off generating values with the properties you desire.

Since I'm not testing anything about the specific values of the three values (all I care about is their relative ordering), I don't see any particular downside to generating three values that are guaranteed to be distinct, and this seemed a straightforward way of doing that. I've now added a comment to this line explaining what I'm doing here, though, since it's not super obvious.

Reply:
If you want to go this route I'd create a strategy for that, so that we can write something like `#[strategy(uniqe_nums(3 /* count */, 0 /* index */, -1000..1000))]` and that we needn't do that in the actual test body.

Reply:
or even more straightforward:

```
fn foo(#[strategy(unique_nums())] [a,b,c]: [i32,3]) { ... }
```

Reply:
That's a nice idea. Does it need to be part of this PR, or can/should that be part of a different issue/PR?

Reply:
I like the idea of it beeing part of this pr but I'll leave it up to you to decide.